### ISSUE TITLE
Unable to add metrics to autoscaling group 

### ISSUE THREAD

**Issue by Sangeli** [terraform-providers/terraform-provider-aws#1098]
I have encountered a bug with enabling autoscaling metrics with an ASG (version 0.9.9). When I run Terraform it shows that it is enabling the metrics but when I go into the AWS console I see it is not true! Here is what Terraform is showing me:

> aws_autoscaling_group.phoenix-asg: Creating...
>   arn:                                "" => "<computed>"
>   availability_zones.#:               "" => "<computed>"
>   default_cooldown:                   "" => "<computed>"
>   desired_capacity:                   "" => "3"
**>   enabled_metrics.#:                  "" => "8"
>   enabled_metrics.119681000:          "" => "GroupStandbyInstances"
>   enabled_metrics.1940933563:         "" => "GroupTotalInstances"
>   enabled_metrics.308948767:          "" => "GroupPendingInstances"
>   enabled_metrics.3267518000:         "" => "GroupTerminatingInstances"
>   enabled_metrics.3394537085:         "" => "GroupDesiredCapacity"
>   enabled_metrics.3551801763:         "" => "GroupInServiceInstances"
>   enabled_metrics.4118539418:         "" => "GroupMinSize"
>   enabled_metrics.4136111317:         "" => "GroupMaxSize"**
>   force_delete:                       "" => "false"
>   health_check_grace_period:          "" => "300"
>   health_check_type:                  "" => "<computed>"
>   launch_configuration:               "" => "phoenix-pciiframe-LC00b6c1e53bc2d57db6c7675da1"
>   load_balancers.#:                   "" => "<computed>"
>   max_size:                           "" => "5"
>   metrics_granularity:                "" => "1Minute"
>   min_size:                           "" => "2"
>   name:                               "" => "<computed>"
>   name_prefix:                        "" => "phoenix-pciiframe-ASG"
>   protect_from_scale_in:              "" => "false"
>   tag.#:                              "" => "3"
>   tag.1492577110.key:                 "" => "Name"
>   tag.1492577110.propagate_at_launch: "" => "true"
>   tag.1492577110.value:               "" => "phoenix-phoenix-pciiframe-ecs-host"
>   tag.3740893210.key:                 "" => "StackId"
>   tag.3740893210.propagate_at_launch: "" => "true"
>   tag.3740893210.value:               "" => "phoenix-pciiframe"
>   tag.3775377659.key:                 "" => "Owner"
>   tag.3775377659.propagate_at_launch: "" => "true"
>   tag.3775377659.value:               "" => "Phoenix Team"
>   target_group_arns.#:                "" => "1"
>   target_group_arns.543432851:        "" => "arn:aws:elasticloadbalancing:us-east-1:717953026772:targetgroup/phoenix-pciiframe-alb-tg/fbf3d0ee5f83dfdd"
>   termination_policies.#:             "" => "1"
>   termination_policies.0:             "" => "OldestInstance"
>   vpc_zone_identifier.#:              "" => "4"
>   vpc_zone_identifier.1300058809:     "" => "subnet-96ec9dbb"
>   vpc_zone_identifier.2680514991:     "" => "subnet-3f270176"
>   vpc_zone_identifier.294314797:      "" => "subnet-a61876fd"
>   vpc_zone_identifier.3365866201:     "" => "subnet-285ae414"
>   wait_for_capacity_timeout:          "" => "10m"
>   wait_for_elb_capacity:              "" => "2"

As far as I can tell it is identical to [this other bug](https://github.com/hashicorp/terraform/issues/7693) but that has been closed for some reason. When I use the AWS console to query the ASG on its properties, I see this:

` "EnabledMetrics": [], `

So this is definitely not a bug of the AWS UI being different than the configuration of the actual resource. Looks like Terraform is somehow not able to add these metrics. I can provide more info if needed.

**Comment by stack72** [comment]
Hi @Sangeli 

I provided an explanation of the error here - https://github.com/hashicorp/terraform/issues/7693#issuecomment-305440549

This is something that currently seems to be broken on the AWS UI

I can confirm that we have acceptance tests that enable and disable metrics for the ASGs and they run nightly. When we query the AWS API, we get the metrics responses back but yet the UI doesn't show the same info as their API

This was what a Terraform plan showed me:

```
% terraform plan                                                                                                                             ✭
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.

aws_launch_configuration.test: Refreshing state... (ID: metrics-test-00905198636fae96a3cc866a52)
aws_autoscaling_group.test: Refreshing state... (ID: metrics-test)
The Terraform execution plan has been generated and is shown below.
Resources are shown in alphabetical order for quick scanning. Green resources
will be created (or destroyed and then created if an existing resource
exists), yellow resources are being changed in-place, and red resources
will be destroyed. Cyan entries are data sources to be read.

Note: You didn't specify an "-out" parameter to save this plan, so when
"apply" is called, Terraform can't guarantee this is what will execute.

~ aws_autoscaling_group.test
    enabled_metrics.#:          "0" => "1"
    enabled_metrics.3394537085: "" => "GroupDesiredCapacity"


Plan: 0 to add, 1 to change, 0 to destroy.
```

This is the API call it is making under the hood:

```
-----------------------------------------------------
aws-provider (internal) 2017/06/01 12:18:46 [DEBUG] [aws-sdk-go] DEBUG: Request autoscaling/EnableMetricsCollection Details:
---[ REQUEST POST-SIGN ]-----------------------------
POST / HTTP/1.1
Host: autoscaling.us-east-1.amazonaws.com
User-Agent: aws-sdk-go/1.8.16 (go1.8.1; darwin; amd64) APN/1.0 HashiCorp/1.0 Terraform/0.9.5-dev
Content-Length: 141
X-Amz-Date: 20170601T091846Z
Accept-Encoding: gzip

Action=EnableMetricsCollection&AutoScalingGroupName=metrics-test&Granularity=1Minute&Metrics.member.1=GroupDesiredCapacity&Version=2011-01-01
-----------------------------------------------------
aws-provider (internal) 2017/06/01 12:18:47 [DEBUG] [aws-sdk-go] DEBUG: Response autoscaling/EnableMetricsCollection Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK
Connection: close
Content-Length: 233
Content-Type: text/xml
Date: Thu, 01 Jun 2017 09:18:47 GMT
X-Amzn-Requestid: 5112ae63-46ab-11e7-b5c5-014b075a0cd5

<EnableMetricsCollectionResponse xmlns="http://autoscaling.amazonaws.com/doc/2011-01-01/">
  <ResponseMetadata>
    <RequestId>5112ae63-46ab-11e7-b5c5-014b075a0cd5</RequestId>
  </ResponseMetadata>
</EnableMetricsCollectionResponse>

-----------------------------------------------------
```
Notice the response is as follows (and has EnabledMetrics included:

```
-----------------------------------------------------
aws-provider (internal) 2017/06/01 12:18:47 [DEBUG] AutoScaling Group describe configuration: {
  AutoScalingGroupNames: ["metrics-test"]
}
aws-provider (internal) 2017/06/01 12:18:47 [DEBUG] [aws-sdk-go] DEBUG: Request autoscaling/DescribeAutoScalingGroups Details:
---[ REQUEST POST-SIGN ]-----------------------------
POST / HTTP/1.1
Host: autoscaling.us-east-1.amazonaws.com
User-Agent: aws-sdk-go/1.8.16 (go1.8.1; darwin; amd64) APN/1.0 HashiCorp/1.0 Terraform/0.9.5-dev
Content-Length: 95
Content-Type: application/x-www-form-urlencoded; charset=utf-8
X-Amz-Date: 20170601T091847Z
Accept-Encoding: gzip

Action=DescribeAutoScalingGroups&AutoScalingGroupNames.member.1=metrics-test&Version=2011-01-01
-----------------------------------------------------
aws-provider (internal) 2017/06/01 12:18:47 [DEBUG] [aws-sdk-go] DEBUG: Response autoscaling/DescribeAutoScalingGroups Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK
Connection: close
Content-Length: 1744
Content-Type: text/xml
Date: Thu, 01 Jun 2017 09:18:47 GMT
X-Amzn-Requestid: 51675dff-46ab-11e7-95ad-59985b1f78db

<DescribeAutoScalingGroupsResponse xmlns="http://autoscaling.amazonaws.com/doc/2011-01-01/">
  <DescribeAutoScalingGroupsResult>
    <AutoScalingGroups>
      <member>
        <HealthCheckType>EC2</HealthCheckType>
        <LoadBalancerNames/>
        <Instances/>
        <TerminationPolicies>
          <member>Default</member>
        </TerminationPolicies>
        <DefaultCooldown>300</DefaultCooldown>
        <EnabledMetrics>
          <member>
            <Metric>GroupDesiredCapacity</Metric>
            <Granularity>1Minute</Granularity>
          </member>
        </EnabledMetrics>
        <AutoScalingGroupARN>arn:aws:autoscaling:us-east-1:187416307283:autoScalingGroup:a5db7cbf-bb86-4960-ab2e-01d93c6d28c3:autoScalingGroupName/metrics-test</AutoScalingGroupARN>
        <MaxSize>0</MaxSize>
        <AvailabilityZones>
          <member>us-east-1a</member>
          <member>us-east-1d</member>
          <member>us-east-1e</member>
        </AvailabilityZones>
        <TargetGroupARNs/>
        <Tags/>
        <LaunchConfigurationName>metrics-test-00905198636fae96a3cc866a52</LaunchConfigurationName>
        <AutoScalingGroupName>metrics-test</AutoScalingGroupName>
        <HealthCheckGracePeriod>300</HealthCheckGracePeriod>
        <NewInstancesProtectedFromScaleIn>false</NewInstancesProtectedFromScaleIn>
        <CreatedTime>2017-06-01T09:17:36.481Z</CreatedTime>
        <MinSize>0</MinSize>
        <SuspendedProcesses/>
        <DesiredCapacity>0</DesiredCapacity>
        <VPCZoneIdentifier/>
      </member>
    </AutoScalingGroups>
  </DescribeAutoScalingGroupsResult>
  <ResponseMetadata>
    <RequestId>51675dff-46ab-11e7-95ad-59985b1f78db</RequestId>
  </ResponseMetadata>
</DescribeAutoScalingGroupsResponse>

-----------------------------------------------------
```

But yet the UI looks as follows:

<img width="890" alt="not-enabled-after" src="https://user-images.githubusercontent.com/227823/28066106-4fbcbfdc-6644-11e7-83eb-71503c370f46.png">

I then checked it via the CLI

```
saws> aws autoscaling describe-auto-scaling-groups --auto-scaling-group-names metrics-test --region us-east-1
{
    "AutoScalingGroups": [
        {
            "AutoScalingGroupARN": "arn:aws:autoscaling:us-east-1:187416307283:autoScalingGroup:a5db7cbf-bb86-4960-ab2e-01d93c6d28c3:autoScalingGroupName/metrics-test",
            "TargetGroupARNs": [],
            "SuspendedProcesses": [],
            "DesiredCapacity": 0,
            "Tags": [],
            "EnabledMetrics": [
                {
                    "Metric": "GroupDesiredCapacity",
                    "Granularity": "1Minute"
                }
            ],
            "LoadBalancerNames": [],
            "AutoScalingGroupName": "metrics-test",
            "DefaultCooldown": 300,
            "MinSize": 0,
            "Instances": [],
            "MaxSize": 0,
            "VPCZoneIdentifier": "",
            "HealthCheckGracePeriod": 300,
            "TerminationPolicies": [
                "Default"
            ],
            "LaunchConfigurationName": "metrics-test-00905198636fae96a3cc866a52",
            "CreatedTime": "2017-06-01T09:17:36.481Z",
            "AvailabilityZones": [
                "us-east-1a",
                "us-east-1d",
                "us-east-1e"
            ],
            "HealthCheckType": "EC2",
            "NewInstancesProtectedFromScaleIn": false
        }
    ]
}
```

So, AFAICT, it's definitely an AWS Console UI bug

Thoughts?

Paul

**Comment by Sangeli** [comment]
> When we query the AWS API, we get the metrics responses back but yet the UI doesn't show the same info as their API

I think that's the key difference between my bug and #7693 that both of us have referenced. I do NOT see enabled metrics in the API response. This is what I see:

> {
    "AutoScalingGroups": [
        {
            "AutoScalingGroupARN": "arn:aws:autoscaling:us-east-1:688497093330:autoScalingGroup:9e812b7a-5f67-49ce-a1ad-8aae8a8cc63f:autoScalingGroupName/phoenix-pciiframe-ASG00ec589bfba0fe31fdcfe28df1", 
            "TargetGroupARNs": [
                "arn:aws:elasticloadbalancing:us-east-1:688497093330:targetgroup/phoenix-pciiframe-alb-tg/ab4bb65a933170bc"
            ], 
            "SuspendedProcesses": [], 
            "DesiredCapacity": 2, 
            "Tags": [
                {
                    "ResourceType": "auto-scaling-group", 
                    "ResourceId": "phoenix-pciiframe-ASG00ec589bfba0fe31fdcfe28df1", 
                    "PropagateAtLaunch": true, 
                    "Value": "Commerce and Conversion Apps", 
                    "Key": "AccountGroupId"
                }, 
                {
                    "ResourceType": "auto-scaling-group", 
                    "ResourceId": "phoenix-pciiframe-ASG00ec589bfba0fe31fdcfe28df1", 
                    "PropagateAtLaunch": true, 
                    "Value": "10.10.200.63741", 
                    "Key": "CostCenter"
                }, 
                {
                    "ResourceType": "auto-scaling-group", 
                    "ResourceId": "phoenix-pciiframe-ASG00ec589bfba0fe31fdcfe28df1", 
                    "PropagateAtLaunch": true, 
                    "Value": "phoenix-phoenix-pciiframe-ecs-host", 
                    "Key": "Name"
                }, 
                {
                    "ResourceType": "auto-scaling-group", 
                    "ResourceId": "phoenix-pciiframe-ASG00ec589bfba0fe31fdcfe28df1", 
                    "PropagateAtLaunch": true, 
                    "Value": "Partha Sarathi  psarathi ", 
                    "Key": "Owner"
                }, 
                {
                    "ResourceType": "auto-scaling-group", 
                    "ResourceId": "phoenix-pciiframe-ASG00ec589bfba0fe31fdcfe28df1", 
                    "PropagateAtLaunch": true, 
                    "Value": "phoenix-pciiframe", 
                    "Key": "StackId"
                }, 
                {
                    "ResourceType": "auto-scaling-group", 
                    "ResourceId": "phoenix-pciiframe-ASG00ec589bfba0fe31fdcfe28df1", 
                    "PropagateAtLaunch": true, 
                    "Value": "Verified", 
                    "Key": "TaggingStatus"
                }
            ], 
            **"EnabledMetrics": [],** 
            "LoadBalancerNames": [], 
            "AutoScalingGroupName": "phoenix-pciiframe-ASG00ec589bfba0fe31fdcfe28df1", 
            "DefaultCooldown": 300, 
            "MinSize": 2, 
            "Instances": [
                {
                    "InstanceId": "i-0aab4e2c8bba31773", 
                    "ProtectedFromScaleIn": false, 
                    "AvailabilityZone": "us-east-1a", 
                    "HealthStatus": "Healthy", 
                    "LifecycleState": "InService"
                }, 
                {
                    "InstanceId": "i-0bd49228af95262d8", 
                    "ProtectedFromScaleIn": false, 
                    "AvailabilityZone": "us-east-1e", 
                    "HealthStatus": "Healthy", 
                    "LifecycleState": "InService"
                }
            ], 
            "MaxSize": 3, 
            "VPCZoneIdentifier": "subnet-6a5ae456,subnet-a12701e8,subnet-39197762,subnet-03ef9e2e", 
            "HealthCheckGracePeriod": 300, 
            "TerminationPolicies": [
                "OldestInstance"
            ], 
            "LaunchConfigurationName": "phoenix-pciiframe-LC00c37d0ce07d48a3e02e712a3c", 
            "CreatedTime": "2017-06-30T22:34:34.735Z", 
            "AvailabilityZones": [
                "us-east-1a", 
                "us-east-1b", 
                "us-east-1d", 
                "us-east-1e"
            ], 
            "HealthCheckType": "EC2", 
            "NewInstancesProtectedFromScaleIn": false
        }
    ]
}

So it looks like the AWS UI and AWS CLI are showing the same thing. Doesn't look like an AWS bug.

**Comment by bflad** [comment]
The fix for this to properly set empty `enabled_metrics` in the state during read (#3899) has been merged into master and will release with v1.13.0 of the AWS provider, likely in about a week.

**Comment by bflad** [comment]
Ahead of schedule, this has been released in [version 1.13.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1130-march-28-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
